---
title: "Oz Single Cells 2018 Informatics Workshop"
author: "Anne Senabouth"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Oz Single Cells 2018 Informatics Workshop}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## 1. Preparing the R Environment
Prior to the workshop, please set up your R environment. Feel free to skip
some steps if you have already done those steps.

### 1.1 R installation
Please follow the R installation instructions [here](https://mirror.aarnet.edu.au/pub/CRAN/).
If you are a Windows user, make sure you install Rtools. Please note the `ascend`
package requires R version >= 3.4.3. The latest version of R version 3.5.1 is 
best.

### 1.2 R programming environment
The workshop will be done in [RStudio](https://www.rstudio.com/products/rstudio/download/),
but feel free to set up your own R workspace.

### 1.3 Package Installations
You will need to install the following packages to run the development version
of `ascend`. Feel free to skip these steps if you already have these packages.

#### 1.3.1 Packages from CRAN
You can use the install.packages() to install the packages described in this 
section. The pcakages you require from this repository are as follows:

1. [devtools](https://cran.r-project.org/web/packages/devtools/index.html): This 
package will be used to load the development version of `ascend`.
2. [tidyverse](https://www.tidyverse.org/): This is a series of R packages 
for data science and visualisation. This will install packages such as dplyr,
ggplot2 and tidyr.
3. [data.table](https://github.com/Rdatatable/data.table/wiki/Installation):
Please follow the instructions on this page for your operating system.

Remaining packages can be installed as follows:

```{r install_cran, eval = FALSE}
# List of packages to install
cran_packages <- c("reshape2", "fields", "ggbeeswarm", "gridExtra", 
                   "dynamicTreeCut", "dendextend", "RColorBrewer",
                   "locfit", "KernSmooth")

# Easy command to install all at once
install.packages(cran_packages)
```

#### 1.3.2 Packages from Bioconductor
Bioconductor is a repository for R packages  related to the analysis and 
comprehension of high-throughput genomic data. It uses a separate set of 
commands for the installation of packages.

##### 1.3.2.1 Setting up Bioconductor
Use the following code to retrieve the latest installer from Bioconductor.

```{r setup_bioconductor, eval = FALSE}
## try http:// if https:// URLs are not supported
source("https://bioconductor.org/biocLite.R")
biocLite()
```

You can then install the Bioconductor packages using `biocLite`.

```{r bioconductor_packages, eval = FALSE}
bioconductor_packages <- c("Biobase", "BiocGenerics", "BiocParallel",
                           "SingleCellExperiment", "scater", "scran")

biocLite(bioconductor_packages)
```

##### 1.3.2.2 Differential expression packages
`ascend` provides wrappers for [DESeq](https://bioconductor.org/packages/release/bioc/html/DESeq.html) 
and [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html), 
so you may choose to add them to your installation. However, we will only be 
using DESeq for the workshop as DESeq2 will require more time than allocated 
for the workshop.

### 1.4 Installing 'ascend' via devtools
As `ascend` is still under development, we will use devtools to install the
package.

```{r install_ascend}
# Load devtools package
library(devtools)

# Use devtools to install the package
install_github("IMB-Computational-Genomics-Lab/ascend", ref = "devel")

# Load the package in R
library(ascend)
```

#### 1.4 Configuring BiocParallel
This package makes extensive use of [BiocParallel](http://bioconductor.org/packages/release/bioc/html/BiocParallel.html), enabling `ascend` to make the most of your computer's hardware. As each system is different, BiocParallel needs to be configured by the user. Here are some example configurations.

#### Unix/Linux/MacOS (Single Machine)
```{r SetupNix, eval = FALSE}
library(BiocParallel)
ncores <- parallel::detectCores() - 1
register(MulticoreParam(workers = ncores, progressbar=TRUE), default = TRUE)
```

#### Windows (Single Machine - Quad-core system)
The following commands allows Windows to parallelise functions via BiocParallel.
Unlike multicore processing in *nix systems, Snow creates additional R sessions 
to export tasks to. This requires additional computational resources to run and 
manage the tasks.

We recomend you bypass this step if your machine has lower specs.

```{r SetupWin, eval = FALSE}
library(BiocParallel)
workers <- 3 # Number of cores on your machine - 1
register(SnowParam(workers = workers, 
                   type = "SOCK", 
                   progressbar = TRUE), default = TRUE)
```

## 2. Preparing the data for the workshop
This vignette will use the unprocessed data from the dataset featured in
the publication [Single cell RNA sequencing of stem cell-derived retinal ganglion cells](https://www.nature.com/articles/sdata201813#data-records) 
by [Daniszewski et al. 2018]().

This dataset consists of 1272 human embryonic stem cell-derived retina ganglion 
cells that have been separated into THY1-positive (Sample 1) and THY1-negative 
(Sample 2) cells via flow cytometry. This protein is a marker for pluripotency.

Please note that this data has been batch equalized with subsampling.

## 2.1 Loading the data into R
We will be bypassing the manual preparation of the data for this workshop. If 
you wish to know more about manually loading the data into R, please refer to 
the `ascend` vignettes.

Instead, we will use the `loadCellRanger` command to load the data into R. This
command parses the filtered data generated by Cell Ranger and assumes 
mitochondrial and ribosomal genes will be used as controls. It also parses 
batch information from the cell identifiers assigned by Cell Ranger's
batch aggregation function. Please note that this batch aggregation function
has also normalised the dataset via the subsampling method described 
by [Zheng et al. 2016](https://www.nature.com/articles/ncomms14049).

```{r load_data}
# Set path to data
data_dir <- "data/iPSC_RGCscRNASeq_PreQC/outs/filtered_gene_bc_matrices_mex/GRCh38p7"
em_set <- loadCellRanger(data_dir)
```

We can review the contents of the EMSet by inputting the variable name of the
EMSet into the console.

```{r review_emset}
em_set
```

The summary shows how many genes are in the EMSet, what genes were set as 
controls and what is kept in the various slots of the EMSet. You may recognise
some of the slot names from somewhere - this is because this version of the
EMSet has inherited from the [SingleCellExperiment](https://bioconductor.org/packages/release/bioc/vignettes/SingleCellExperiment/inst/doc/intro.html#the-singlecellexperiment-class) superclass.

The EMSet differs from the SingleCellExperiment class in which it contains 
seperate slots for cell and gene-related metadata and additional slots for
cluster analysis and log information. Also, the EMSet generates a set of
quality control metrics based on the count matrix upon creation, and will
update the metrics everytime this matrix is changed.

We can review the cell information that is stored in the colInfo slot as follows:

```{r review_colinfo}
# Cell-related metadata
colInfo(em_set)

# Cell-related metrics
colData(em_set)

# Gene-related metadata
rowInfo(em_set)

# Gene-related metrics
rowData(em_set)
```

## 3. Quality Control and Filtering
This workflow is similar to other scRNA-seq filtering workflows available
through packages such as `seurat` and `scater`. You can read more about quality 
control through these links:

1. [Analyzing single-cell RNA-seq data containing UMI counts](https://bioconductor.org/packages/release/workflows/vignettes/simpleSingleCell/inst/doc/work-2-umis.html) by Lun, McCarthy & Marioni 2018
2. [Cleaning the Expression Matrix](https://hemberg-lab.github.io/scRNA.seq.course/cleaning-the-expression-matrix.html#exprs-qc) by Kiselex, Andrews, McCarthy, Buttner & Hemberg 2018

### 3.1 Assessing quality of the dataset
First, we will have a look at the quality of the dataset. Some quality control
metrics were generated when the EMSet was created; they are stored in the
`rowData` (metrics for genes) and `colData` (metrics for cells) slots. We can
visualise these values with quality control plots.

We can generate all the plots we will need for this step with with the
plotGeneralQC function.

```{r raw_qc_func, fig.show = "hide", results = "hide", include = FALSE}
qc_plots <- plotGeneralQC(em_set)
```

```{r raw_qc_show, eval = FALSE}
qc_plots <- plotGeneralQC(em_set)
```


The two plots below represent metrics related to the total number of UMIs 
mapped to a cell.

The first plot is a barplot that depicts the library size of each cell and is
coloured by the batch. This plot lets us quickly review the distribution of
library sizes and if there are any significant batch effects in the dataset.

The second plot is a histogram of library sizes, with library size on the x-axis
and number of cells with the library size on the y-axis. Please note that this 
dataset comprises of cells that have already been filtered from the background
based on a minimum threshold value for the number of UMIs present in the cell.
There are some cells with significantly larger numbers of UMIs - they may be
doublets where more than one cell ended up in a droplet. We will need to 
remove these cells.

```{r libsize_qc, fig.width=6, fig.height=2.5}
library(gridExtra)
grid.arrange(qc_plots$libsize_barplot, qc_plots$libsize_histogram, ncol = 1)
```

We also want to review the number of genes being expressed in each cell, and
what proportion of UMIs are mapped to top gene expressors. The first plot below
represents the number of cells expressing a number of genes. The number of
expressed genes will differ based on the cell type and other factors such 
as cell cycle phase.

The second plot represents the proportion of genes mapped to the top 500 most
expressed genes (y-axis) per sample, with each point representing the percentage
of expression contributed by the 100 most-expressed genes. 

```{r gene_qc, fig.width=6, fig.height=4}
grid.arrange(qc_plots$ngenes_hist, qc_plots$topgenes_violin, ncol = 1)
```

Some cells appear to be dominated by the expression of a small subset of genes. 
These genes may be mitochondrial and ribosomal genes. 

The plots below show the proportion of mitochondrial and ribosomal gene
expression to total expression. Some cells have a higher proportion of ribosomal
gene expression that can indicate they are of low quality, and should be 
removed from the dataset.

```{r control_qc, fig.width=6, fig.height = 3}
grid.arrange(qc_plots$control_hists$Mt, qc_plots$control_hists$Rb, ncol = 2)
```

```{r control_qc2, fig.width = 6, fig.height = 6}
grid.arrange(qc_plots$control_violins$Mt, qc_plots$control_violins$Rb, ncol = 1)
```

### 3.2 Batch normalisation
Before we start filtering, we need to normalise the counts between the two 
batches to account for technical variations. The read counts have already been 
normalised already via the Cell Ranger pipeline's subsampling method. 
This batch normalisation method scales UMI counts between batches.

```{r batchNorm}
em_set <- normaliseBatches(em_set)
```

### 3.3 Filtering low quality cells from the dataset
There are three filtering functions in the `ascend` package. 

1. `filterByOutliers` removes outliers in terms of library size and control
expression based on median absolute deviation (MAD).
2. `filterByControl` removes cells where genes in the specified control group
account for more than the specified percentage of expression.
3. `filterLowAbundanceGenes` removes genes that are expressed in less than
the specified percentage of cells.

```{r filter_data}
# Filter library size and controls by MAD
filtered_set <- filterByOutliers(em_set, 
                                 cell.threshold = 3, 
                                 control.threshold = 3)

# Filter out cells based on proportion of control expression
filtered_set <- filterByControl(filtered_set, control = "Mt", pct.threshold = 20)
filtered_set <- filterByControl(filtered_set, control = "Rb", pct.threshold = 50)
filtered_set <- filterLowAbundanceGenes(filtered_set, pct.threshold = 1)
```

We can review which cells and genes were removed from the dataset by reviewing
the progress log. As thousands of low abundance genes were removed, we'll look
at the summary of the log.

```{r review_log}
str(progressLog(filtered_set))
```

We can also review the impact of QC on the dataset by regenerating the QC plots.

```{r review_qc_plots_run, fig.show = "hide", results = "hide", include = FALSE}
filtered_qc_plots <- plotGeneralQC(filtered_set)
```

```{r review_qc_plots, eval = FALSE}
filtered_qc_plots <- plotGeneralQC(filtered_set)
```

```{r display_review_qc, fig.width = 6, fig.height = 7}
grid.arrange(filtered_qc_plots$libsize_histogram, filtered_qc_plots$ngenes_hist, 
             filtered_qc_plots$control_hists$Mt, filtered_qc_plots$control_hists$Rb, 
             filtered_qc_plots$control_violins$Mt, filtered_qc_plots$control_violins$Rb,
             ncol = 2)
```

### 4. Cell-cell normalisation
`ascend` offers a normalisation method based on Relative Log Expression. This 
is the method used by DESeq to normalise counts, but has been adapated for 
zero-inflated data.

In this method, each cell is considered as one library and assumes that most 
genes are not differentially expressed. It uses gene expression values higher 
than 0 to calculate the geometric means of a gene. The geometric mean is the
mean of the expression of the gene across all cells (for cells where the gene is 
detected). Each gene has one geometric mean value for all cell. For each cell,
the gene expression values are divided by the geometric means to get one 
normalisation factor for a gene in that cell. The median of all the 
normalisation factors for all genes in that cell is the normalisation factor for 
the cell. 

Post RLE normalisation, a gene with 0 expression still has 0 expression. A gene 
with expression higher than 0 will have an expression value equal the raw expression 
divided by the calculated normalization factor for the cell.

```{r NormaliseByRLE}
norm_set <- normaliseByRLE(filtered_set)
```

Original counts are retained in the EMSet. Normalised counts and log-transformed
normalised counts are stored in the EMSet as well, accessible via the 
`normcounts` and `logcounts` accessors from the SingleCellExperiment class.

```{r reviewnorm}
# Raw counts
counts(norm_set)[1:5,1:5]

# Normalised counts
normcounts(norm_set)[1:5,1:5]

# Log2(counts) + 1
logcounts(norm_set)[1:5, 1:5]
```

We can generate some plots to review the impact of normalisation on the dataset.

```{r generateNormPlots}
norm_qc <- plotNormQC(norm_set)
```

This set of plots depict the library sizes of cells prior to and after 
RLE normalisation.

```{r norm_libsize_plots, fig.width = 6, fig.height = 4}
grid.arrange(norm_qc$libsize_histograms$count, 
             norm_qc$libsize_histograms$normcount, ncol = 1)
```

The following plots depict the impact of RLE normalisation on GAPDH counts.

```{r norm_gapdh, fig.width = 6, fig.height = 4}
grid.arrange(norm_qc$sampled_genes$GAPDH$counts,
             norm_qc$sampled_genes$GAPDH$normcounts, ncol = 1)
```

The following plots depict the impact of RLE normalisation on MALAT1 counts.

```{r norm_malat1, fig.width = 6, fig.height = 4}
grid.arrange(norm_qc$sampled_genes$MALAT1$counts,
             norm_qc$sampled_genes$MALAT1$normcounts, ncol = 1)
```

The following plots depict the impact of RLE normalisation on all gene counts
from 100 randomly-selected cells.

```{r norm_genescatter, fig.width = 6, fig.height = 4}
grid.arrange(norm_qc$sampled_cell_gene_expression$counts,
             norm_qc$sampled_cell_gene_expression$normcounts, ncol = 1)
```

We have also included a wrapper for `scran`'s deconvolution method.

## 5. Control removal
We can review the top expressors by examining the top gene expression boxplot.

```{r, fig.width = 6, fig.height = 5}
filtered_qc_plots$topgenes_boxplot
```

As mitochondrial and ribosomal genes still dominate top expression after filtering,
we should remove them from the dataset so we can observe genes of interest.

```{r exclude_controls}
# Exclude mitochondrial and ribosomal genes
norm_set <- excludeControl(norm_set, control = c("Mt", "Rb"))
```

```{r plotTopGenesNoControls, fig.width = 6, fig.height = 5}
print(plotTopGenesBoxplot(norm_set, n = 20))
```

## 6. Dimensionality reduction
Dimensionality reduction is used to reduce the number of variables we have to
assess which will reduce noise and speed up subesequent analyses. It is also 
used to visualise the data in 2 or 3 dimensional space.

We will transform our data with Principal Component Analysis (PCA).

```{r reducePCA}
pca_set <- runPCA(norm_set, ngenes = 1500, scaling = TRUE)
reducedDim(pca_set, "PCA")[1:5,1:5]
```

We can observe how much variance each principal component contributes to a 
dataset using a scree plot. The first PC will contribute the most variance
to a dataset, but the number of PCs that do contribute variance will differ.

```{r plotPCAVariance, fig.width = 6, fig.height = 3}
plotPCAVariance(pca_set, n = 50)
```

As previously mentioned, these values can also be used to represent the data
in two dimensions.

```{r plotPCABatch, fig.width = 6, fig.height = 5.5}
plotPCA(pca_set, PCX = 1, PCY = 2, group = "batch")
```

## 7. Clustering
### 7.1 Using runCORE
The `runCORE` function generates a distance matrix based on the input and from 
this, builds a dendrogram. This dendrogram is then cut with the `dynamicTreeCut` 
algorithm to select clusters from the dendrogram based on the shape and size of 
the branches. The cut is then repeated over a range of heights ranging from 0
to 1. The clusters generated by the most robust height - defined as the point 
where cluster numbers stabilise, is chosen as the optimal clustering result.

```{r RunCORE}
clustered_set <- runCORE(pca_set, 
                         conservative = FALSE, 
                         remove_outlier = TRUE,
                         nres = 40)
```

`runCORE` generates a series of clustering-related objects, that can be
accessed as follows:

```{r reviewCORE}
cluster_analysis <- clusterAnalysis(clustered_set)

# Number of clusters
cluster_analysis$nClusters

# Rand Index values
cluster_analysis$keyStats
```

### 7.2 Visualising clustering results
The `plotStabilityDendro` and `plotStability` functions allows us to review the
clustering process by showing other clustering results and their corresponding
Rand index.

```{r stability_plot, fig.width = 5, fig.height = 6}
plotStabilityDendro(clustered_set)
```

```{r plotStability, fig.width = 6, fig.height = 5.5}
plotStability(clustered_set)
```

`plotDendrogram` generates a cluster-labelled dendrogram that also displays
the size of each cluster.

```{r plotDendrogram,  fig.width = 6, fig.height = 4}
plotDendrogram(clustered_set)
```

The clustering information can also be shown on plots as the information
has been added to colInfo.

```{r plotTSNE}
# Reduce dimensionality with t-SNE
clustered_set <- runTSNE(clustered_set, PCA = TRUE, 
                        dimensions = 2, seed = 1, 
                        perplexity = 30, theta = 0.5)

# Generate a t-SNE plot
tsne_plot <- plotTSNE(clustered_set, Dim1 = 1, Dim2 = 2, group = "cluster")

# Generate a PCA plot
pca_plot <- plotPCA(clustered_set, PCX=1, PCY=2, group = "cluster")

# Generate MDS plot
mds_plot <- plotMDS(clustered_set, Dim1 = 1, Dim2 = 2, group = "cluster")
```
```{r tsne_plot, fig.width = 6, fig.height = 5}
tsne_plot
```

```{r cluster_pca_plot, fig.width = 6, fig.height = 5}
pca_plot
```

```{r mds_plot, fig.width = 6, fig.height = 5}
mds_plot
```

## Differential expression
For differential expression analysis, `ascend` offers wrappers for the DESeq and 
DESeq2 packages. However, they are not suited for some datasets and systems.

Let's quickly review our dataset.
```{r review_dataset}
col_info <- colInfo(clustered_set)
# Batch numbers
print("Cells per batch")
table(col_info$batch)

# Cluster numbers
print("Cells per cluster")
table(col_info$cluster)
```

This dataset is unbalanced as the two batches and three clusters
vary significantly in size. These are not accounted for in some models used for 
differential expression. 

We decided to implement a method described by McDavid, Finak, Chattopadyay et al.
2013 [https://academic.oup.com/bioinformatics/article-lookup/doi/10.1093/bioinformatics/bts714] that accounts for the bimodal nature of scRNA-seq
data for differential expression analysis in `ascend`. 

Firstly, genes are either on or off and can be represented as a
discrete model. Secondly, if genes are being expressed, it can be represented
with a continuous model. The discrete and continuous models are incorporated
into a combined Likelihood Ratio Test (LRT). Genes with no variance are assumed
to only have a discrete model.

This method performs analysis on a gene-gene basis and thus is not affected by
differences in sample sizes. However, as this method only studies gene-gene
differences, ....

```{r LRT}
# Run combined LRT
cluster1_vs_others <- runDiffExpression(clustered_set, group = "cluster",
                                        condition.a = 1, condition.b = c(2, 3),
                                        subsampling = TRUE, ngenes = 1500)
cluster2_vs_others <- runDiffExpression(clustered_set, group = "cluster",
                                        condition.a = 2, condition.b = c(1, 3),
                                        subsampling = TRUE, ngenes = 1500)
cluster3_vs_others <- runDiffExpression(clustered_set, group = "cluster",
                                        condition.a = 3, condition.b = c(1, 2),
                                        subsampling = TRUE, ngenes = 1500)
```

```{r plotLRT}
# Generate volcano plots
cluster1_volcano <- plotVolcano(cluster1_vs_others, l2fc = 2, 
                                threshold = 5e-3, 
                                labels = TRUE,
                                label.size = 3,
                                check.overlap = TRUE)

cluster2_volcano <- plotVolcano(cluster2_vs_others, l2fc = 2, 
                                threshold = 5e-3, 
                                labels = TRUE,
                                label.size = 3,
                                check.overlap = TRUE)

cluster3_volcano <- plotVolcano(cluster3_vs_others, l2fc = 2, 
                                threshold = 5e-3, 
                                labels = TRUE,
                                label.size = 3,
                                check.overlap = TRUE)

library(ggplot2)
cluster1_volcano <- cluster1_volcano + ggtitle("Cluster 1 vs Cluster 2 and 3")
cluster2_volcano <- cluster2_volcano + ggtitle("Cluster 2 vs Cluster 1 and 3")
cluster3_volcano <- cluster3_volcano + ggtitle("Cluster 3 vs Cluster 1 and 2")
```

```{r cluster1_volcano, fig.width = 6, fig.height = 5}
cluster1_volcano
```

```{r cluster2_volcano, fig.width = 6, fig.height = 5}
cluster2_volcano
```

```{r cluster3_volcano, fig.width = 6, fig.height = 5}
cluster3_volcano
```

## Branching out with other R packages
In this tutorial, we covered the main `ascend` workflow that completes the 
following the tasks:

1. Quality control
2. Filtering
3. Normalisation
4. Dimensionality reduction
5. Clustering
6. Differential expression

As the package focuses on so many steps, it does not cover the individual steps
in as great detail as other packages. Instead, we have chosen methods that are 
fast and effective. If you require more detailed analysis, you may export
the data stored in the EMSet for use with other packages.

```{r conversion}
sce_object <- EMSet2SCE(clustered_set)

# Extract data from EMSet
# Expression matrix
count_matrix <- counts(clustered_set)
normalised_matrix <- normcounts(clustered_set)
log_matrix <- logcounts(clustered_set)

```

### Package wrappers
We have created wrappers for specific tasks in `scran`, `DESeq` and `DESeq2`.
They provide alternatives for specific stages in the `ascend` workflow and 
provide more in-depth analysis.

#### scran

##### scran normalisation by deconvolution
```{r scran_normalisation}
scran_normalised <- scranNormalise(filtered_set, quickCluster = FALSE, min_mean = 1e-05)
```

### DESeq
```{r DESeq}
cluster1_vs_others <- runDESeq(clustered_set, group = "cluster",
                               condition.a = 1, condition.b = c(2, 3),
                               ngenes = 1500)
```

### DESeq2
DESeq2 requires more time than allowed for this workshop, so we will not run it.
```{r DESeq2, eval = FALSE}
cluster1_vs_others <- runDESeq2(clustered_set, group = "cluster",
                               condition.a = 1, condition.b = c(2, 3),
                               ngenes = 1500)
```